File: backend/src/ZooSanMarino.Infrastructure/Services/CuadreAlimentoEngordeService.cs
// src/ZooSanMarino.Infrastructure/Services/CuadreAlimentoEngordeService.cs
// Lee el invariante del alimento de engorde desde fn_cuadre_alimento_engorde y lo clasifica con
// CuadreAlimentoEngordeCalculos (puro y testeado).
//
// El descuadre de jul-2026 lo detectó un humano de operación semanas después de producirse: nada en el
// sistema lo verificaba. Este service es lo que faltaba.
using Microsoft.EntityFrameworkCore;
using ZooSanMarino.Application.Calculos;
using ZooSanMarino.Application.DTOs;
using ZooSanMarino.Application.Interfaces;
using ZooSanMarino.Domain.Entities;
using ZooSanMarino.Infrastructure.Persistence;

namespace ZooSanMarino.Infrastructure.Services;

public partial class CuadreAlimentoEngordeService : ICuadreAlimentoEngordeService
{
    private readonly ZooSanMarinoContext _db;
    private readonly ICurrentUser? _current;
    private readonly ICompanyResolver _companyResolver;

    public CuadreAlimentoEngordeService(
        ZooSanMarinoContext db, ICurrentUser? current, ICompanyResolver companyResolver)
    {
        _db = db;
        _current = current;
        _companyResolver = companyResolver;
    }

    /// <summary>Proyección cruda de la función SQL. Nombres en snake_case, como exige EF con SqlQueryRaw.</summary>
    private sealed class FilaCruda
    {
        public int      company_id            { get; set; }
        public string   empresa               { get; set; } = "";
        public int      granja_id             { get; set; }
        public string   granja                { get; set; } = "";
        public string   nucleo_id             { get; set; } = "";
        public string   galpon_id             { get; set; } = "";
        public int      lote_ave_engorde_id   { get; set; }
        public string   lote_nombre           { get; set; } = "";
        public string   estado_operativo_lote { get; set; } = "";
        public DateTime ultimo_seguimiento    { get; set; }
        public double   saldo_tabla_kg        { get; set; }
        public double   mov_post_kg           { get; set; }
        public double   stock_kg              { get; set; }
        public double   esperado_kg           { get; set; }
        public double   descuadre_kg          { get; set; }
        public int      filas_negativas       { get; set; }
    }

    /// <summary>
    /// Empresa efectiva: la activa del contexto manda sobre el claim, igual que
    /// <c>InventarioGestionService.GetEffectiveCompanyIdAsync</c>.
    /// <para>
    /// Fail-closed: devuelve 0 si no se puede resolver. Un cuadre que mezclara empresas sería peor que
    /// no tenerlo (mismo criterio que <c>InventarioCatalogoScopeCalculos</c>).
    /// </para>
    /// </summary>
    private async Task<int> ResolverCompanyIdAsync()
    {
        int companyId = 0;
        if (_current is not null)
        {
            if (!string.IsNullOrWhiteSpace(_current.ActiveCompanyName))
            {
                var porNombre = await _companyResolver.GetCompanyIdByNameAsync(_current.ActiveCompanyName.Trim());
                if (porNombre.HasValue) companyId = porNombre.Value;
            }
            if (companyId <= 0 && _current.CompanyId > 0)
                companyId = _current.CompanyId;
        }
        return companyId;
    }

    public async Task<CuadreAlimentoEngordeDto> ObtenerAsync(
        bool soloConProblemas = false, CancellationToken ct = default)
    {
        var companyId = await ResolverCompanyIdAsync();
        if (companyId <= 0)
            return new CuadreAlimentoEngordeDto(0, 0, 0, 0, 0m, []);

        var crudas = await _db.Database
            .SqlQueryRaw<FilaCruda>("SELECT * FROM fn_cuadre_alimento_engorde({0}::int)", companyId)
            .ToListAsync(ct);

        // Kilos SEPARADOS y todavia sin aplicar, por ubicacion. Con doble validacion el consumo de un
        // registro pendiente ya esta dentro de `saldo_tabla` (ninguna fn mira `validado`) pero todavia
        // no salio del inventario: sin restarlo del stock, cada pendiente se reportaba como descuadre.
        // Con el flag apagado no hay reservas ACTIVAS y este diccionario queda vacio.
        var reservado = (await _db.SeguimientoReservaAlimento.AsNoTracking()
                .Where(x => x.CompanyId == companyId && x.Estado == EstadoReservaSeguimiento.Activa)
                .GroupBy(x => new { x.FarmId, x.NucleoId, x.GalponId })
                .Select(g => new { g.Key.FarmId, g.Key.NucleoId, g.Key.GalponId, Kg = g.Sum(x => x.CantidadKg) })
                .ToListAsync(ct))
            .ToDictionary(
                x => (x.FarmId, (x.NucleoId ?? "").Trim(), (x.GalponId ?? "").Trim()),
                x => x.Kg);

        // Correcciones manuales de stock DENTRO del ciclo activo. La fn diaria no las ve (se espejan
        // como INV_OTRO y ninguna de sus 5 CTE lee ese tipo), así que son la causa más frecuente de un
        // galpón descuadrado: en ItalcolPanama explican 42.494 de 54.795 kg (17ago26). Las ANTERIORES
        // al ciclo no se cuentan: ya las tomó la apertura al arrancar. Ojo — un ajuste NO descuadra por
        // sí solo (ItalcolEcuador tiene 5 galpones con ajustes dentro del ciclo y los 36 cuadran), así
        // que esto es una PISTA, no un veredicto: por eso se informa y no se resta del descuadre.
        var loteIds = crudas.Select(r => r.lote_ave_engorde_id).Distinct().ToList();

        var inicioPorLote = (await _db.SeguimientoDiarioAvesEngorde.AsNoTracking()
                .Where(s => loteIds.Contains(s.LoteAveEngordeId))
                .GroupBy(s => s.LoteAveEngordeId)
                .Select(g => new { LoteId = g.Key, Inicio = g.Min(s => s.Fecha) })
                .ToListAsync(ct))
            .ToDictionary(x => x.LoteId, x => x.Inicio.Date);

        var granjaIds = crudas.Select(r => r.granja_id).Distinct().ToList();

        var ajustes = await _db.InventarioGestionMovimientos.AsNoTracking()
            .Where(m => m.CompanyId == companyId
                     && granjaIds.Contains(m.FarmId)
                     && (m.MovementType == "AjusteStock" || m.MovementType == "EliminacionStock"))
            .Select(m => new
            {
                m.FarmId,
                Nucleo = m.NucleoId == null ? "" : m.NucleoId.Trim(),
                Galpon = m.GalponId == null ? "" : m.GalponId.Trim(),
                m.Quantity,
                Fecha = m.CreatedAt
            })
            .ToListAsync(ct);

        var filas = crudas.Select(r =>
        {
            reservado.TryGetValue((r.granja_id, (r.nucleo_id ?? "").Trim(), (r.galpon_id ?? "").Trim()),
                out var reservadoKg);
            var descuadre = CuadreAlimentoEngordeCalculos.DescuadreAjustadoPorReservas(
                (decimal)r.descuadre_kg, reservadoKg);

            var nucleo = (r.nucleo_id ?? "").Trim();
            var galpon = (r.galpon_id ?? "").Trim();
            var desde  = inicioPorLote.TryGetValue(r.lote_ave_engorde_id, out var ini)
                             ? ini
                             : r.ultimo_seguimiento.Date;

            var delGalpon = ajustes.Where(a => a.FarmId == r.granja_id
                                            && a.Nucleo == nucleo
                                            && a.Galpon == galpon
                                            && a.Fecha.UtcDateTime.Date >= desde).ToList();

            // Magnitud movida a mano, sin signo: el `quantity` del ajuste ya es el delta contra el
            // saldo anterior. No se resta del descuadre — un ajuste es una corrección real que alguien
            // tiene que decidir, no ruido de medición como sí lo era la reserva de la doble validación.
            var ajustesKg = delGalpon.Sum(a => Math.Abs(a.Quantity));

            return new CuadreAlimentoEngordeFilaDto(
                r.company_id, r.empresa, r.granja_id, r.granja, r.nucleo_id, r.galpon_id,
                r.lote_ave_engorde_id, r.lote_nombre, r.estado_operativo_lote, r.ultimo_seguimiento,
                (decimal)r.saldo_tabla_kg, (decimal)r.mov_post_kg, (decimal)r.stock_kg,
                (decimal)r.esperado_kg, descuadre, r.filas_negativas,
                CuadreAlimentoEngordeCalculos.Clasificar(descuadre, r.filas_negativas),
                CuadreAlimentoEngordeCalculos.DescribirConAjustes(
                    descuadre, r.filas_negativas, ajustesKg, delGalpon.Count),
                ajustesKg,
                delGalpon.Count);
        }).ToList();

        // El resumen se calcula SIEMPRE sobre el total; el filtro solo recorta el detalle, para que
        // «0 descuadrados de 35» siga siendo legible aunque se pida solo lo problemático.
        var resumen = new CuadreAlimentoEngordeDto(
            TotalGalpones:    filas.Count,
            Cuadran:          filas.Count(f => f.Estado == EstadoCuadreAlimento.Ok),
            Descuadrados:     filas.Count(f => f.Estado == EstadoCuadreAlimento.Descuadrado),
            ConSaldoNegativo: filas.Count(f => f.Estado == EstadoCuadreAlimento.SaldoNegativo),
            KgErrorAbsoluto:  filas.Sum(f => Math.Abs(f.DescuadreKg)),
            Galpones:         soloConProblemas
                                  ? filas.Where(f => f.Estado != EstadoCuadreAlimento.Ok).ToList()
                                  : filas);

        return resumen;
    }
}

File: frontend/src/app/features/gestion-inventario/services/cuadre-alimento-engorde.service.ts
// src/app/features/gestion-inventario/services/cuadre-alimento-engorde.service.ts
//
// Los dos detectores del alimento de engorde:
//  · el CUADRE por galpón (saldo del ciclo activo == stock − movimientos posteriores), que existía en
//    el backend desde jul-2026 y no tenía una sola pantalla que lo mostrara;
//  · el señalamiento de la anomalía R2: lotes que se liquidaron dejando alimento en el galpón.
import { Injectable, inject } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from '../../../../environments/environment';

/** Veredicto del cuadre de un galpón. El API serializa el enum como número. */
export enum EstadoCuadreAlimento {
  Ok = 0,
  SaldoNegativo = 1,
  Descuadrado = 2
}

/** Veredicto de un lote liquidado con saldo de alimento. */
export enum EstadoAlimentoLiquidado {
  Trasladado = 0,
  PendienteEnGalpon = 1,
  SinRespaldoFisico = 2
}

export interface CuadreAlimentoEngordeFilaDto {
  companyId: number;
  empresa: string;
  granjaId: number;
  granja: string;
  nucleoId: string;
  galponId: string;
  loteAveEngordeId: number;
  loteNombre: string;
  estadoOperativoLote: string;
  ultimoSeguimiento: string;
  saldoTablaKg: number;
  movPostKg: number;
  stockKg: number;
  esperadoKg: number;
  descuadreKg: number;
  filasNegativas: number;
  estado: EstadoCuadreAlimento;
  detalle: string;
  /**
   * Kilos corregidos a mano en el stock (`AjusteStock`/`EliminacionStock`) dentro del ciclo activo.
   * La tabla diaria no los ve, así que cuando el galpón no cuadra suelen ser la causa.
   */
  ajustesManualesKg: number;
  ajustesManualesCount: number;
  /**
   * Kilos que la doble validacion tiene SEPARADOS y todavia sin aplicar en esta ubicacion.
   *
   * 🔴 Es la razon por la que `descuadreKg` NO es `saldoTablaKg - (stockKg - movPostKg)`: el
   * descuadre que publica el backend ya viene corregido por este numero. Quien recalcule el
   * invariante en la pantalla tiene que restarlo del stock o le va a dar distinto que la fila.
   * Con el flag de doble validacion apagado es siempre 0.
   */
  reservadoActivoKg: number;
}

export interface CuadreAlimentoEngordeDto {
  totalGalpones: number;
  cuadran: number;
  descuadrados: number;
  conSaldoNegativo: number;
  kgErrorAbsoluto: number;
  galpones: CuadreAlimentoEngordeFilaDto[];
}

export interface AnomaliaAlimentoLiquidadoFilaDto {
  companyId: number;
  granjaId: number;
  granja: string;
  nucleoId: string;
  galponId: string;
  loteAveEngordeId: number;
  loteNombre: string;
  liquidadoAt: string;
  ultimoSeguimiento: string | null;
  saldoCongeladoKg: number;
  salidasPostKg: number;
  stockGalponKg: number;
  kgSinTrasladar: number;
  kgSinRespaldo: number;
  estado: EstadoAlimentoLiquidado;
  detalle: string;
  loteSiguienteId: number | null;
  loteSiguienteNombre: string | null;
  loteSiguienteEncaset: string | null;
}

export interface AnomaliaAlimentoLiquidadoDto {
  totalLiquidados: number;
  conSaldo: number;
  sinDatoCongelado: number;
  pendientesEnGalpon: number;
  sinRespaldoFisico: number;
  kgSinTrasladar: number;
  kgSinRespaldo: number;
  lotes: AnomaliaAlimentoLiquidadoFilaDto[];
}

@Injectable({ providedIn: 'root' })
export class CuadreAlimentoEngordeService {
  private readonly http = inject(HttpClient);
  private readonly api = `${environment.apiUrl}/CuadreAlimentoEngorde`;

  /** Cuadre de todos los galpones de la empresa activa. */
  obtenerCuadre(soloConProblemas = false): Observable<CuadreAlimentoEngordeDto> {
    const params = new HttpParams().set('soloConProblemas', soloConProblemas);
    return this.http.get<CuadreAlimentoEngordeDto>(this.api, { params });
  }

  /** Lotes ya liquidados que congelaron su liquidación con alimento en el galpón. */
  obtenerLiquidadosConAlimento(soloAnomalias = false): Observable<AnomaliaAlimentoLiquidadoDto> {
    const params = new HttpParams().set('soloAnomalias', soloAnomalias);
    return this.http.get<AnomaliaAlimentoLiquidadoDto>(`${this.api}/liquidados-con-alimento`, { params });
  }

  /**
   * Cierra el descuadre de un galpón declarando los kilos que realmente hay.
   *
   * El backend deriva qué escribir de cada lado: si sobra stock corrige el inventario, si sobra
   * tabla corrige la tabla diaria. La pantalla no elige el lado — quien está frente al galpón sabe
   * cuánto alimento hay, no dónde está el error.
   */
  cuadrarGalpon(req: CuadrarGalponAlimentoRequest): Observable<CuadrarGalponAlimentoResultDto> {
    return this.http.post<CuadrarGalponAlimentoResultDto>(`${this.api}/cuadrar-galpon`, req);
  }
}

export interface CuadrarGalponAlimentoRequest {
  loteAveEngordeId: number;
  /** Ítem de alimento sobre el que se registra la corrección. */
  itemInventarioEcuadorId: number;
  /** Kilos que hay físicamente en el galpón. Cero es válido. */
  kilosRealesKg: number;
  /** Obligatorio (mínimo 10 caracteres): queda en la auditoría del movimiento. */
  motivo: string;
}

export interface CuadrarGalponAlimentoResultDto {
  granja: string;
  nucleoId: string;
  galponId: string;
  loteNombre: string;
  saldoTablaAntesKg: number;
  stockAntesKg: number;
  movPostKg: number;
  kilosRealesKg: number;
  /** Kilos escritos en el inventario. 0 = el inventario ya estaba bien. */
  deltaStockKg: number;
  /** Kilos escritos en la tabla diaria. 0 = la tabla ya estaba bien. */
  deltaTablaKg: number;
  descuadreAntesKg: number;
  descuadreDespuesKg: number;
  resumen: string;
}
